### PR TITLE
Assignment 8 : mmwoodfo

### PR DESCRIPTION
**Pull request by mmwoodfo**

### CODE REVIEW

**Comment by ashermancinelli** [comment]
Note that you still have the original tests directory tracked by git even though you have a new one!

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -33,6 +34,28 @@ public IEnumerable<string> CsvRows
...
+        //--------------- 2 : Get Unique Sorted List Of States Given CSV Rows ----------------//
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            var states =
+                CsvRows
+                .Select(row => ParsePerson(row))
+                .Select(p => p.Address.State);
+
+            return states.OrderBy(s => s).Distinct();
+        }
```
Comment (ashermancinelli) [inline Assignment/SampleData.cs]:
```c#
        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() =>
                CsvRows
                .Select(row => ParsePerson(row))
                .Select(p => p.Address.State)
                .OrderBy(s => s)
                .Distinct();
```

Nothing technically different but this feels easier to understand to me!

Comment (mmwoodfo) [reply Assignment/SampleData.cs]: Ah, thank you. I took heavy inspiration from your PR to create a separate parse person method, but tried to piece the rest of the methods together using the Microsoft documentation which is probably why turned out like that, haha. 

I'll go through and refactor this method, I used this pattern on most of my methods so I'll probably clean those up too.

Comment (ashermancinelli) [reply Assignment/SampleData.cs]: Haha fair enough. The microsoft docs seem to me to be easier to understand but maybe not the cleanest implementation.

File: AssignmentTests/SampleDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+using System.Linq;
+
+namespace Assignment.Tests
+{
+    [TestClass()]
+    public class SampleDataTests
+    {
+        private string _FilePath = "TestData.csv";
+
+        //----------------- SAMPLE DATA TEST ----------------//
```
Comment (AustinHall7798) [inline AssignmentTests/SampleDataTests.cs]: I like the way you used the comments to organize your tests. It makes it very readable and I'll use this in my own future test classes!

Comment (mmwoodfo) [reply AssignmentTests/SampleDataTests.cs]: Thank you! I know Mark originally had everything labeled 1-6, so I wanted to keep that labeling to make it easier for grading and for me when I navigate the code.

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,106 @@
...
+        //--------------- 3 : Get Aggregated Sorted List Of States Using CSV Rows ----------------//
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            IEnumerable<string> uniqueSortedListOfStates = GetUniqueSortedListOfStatesGivenCsvRows();
+
+            string[] states = uniqueSortedListOfStates.ToArray();
+
+            return string.Join(",", states);
+        }
```
Comment (GamerBah) [inline Assignment/SampleData.cs]: Using `string.Join()` here is nice, but [LINQ also has a handy thing for enumerables](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.aggregate?view=netcore-3.0) -- `enumerable.Aggregate()`

Using that, you can have some pretty short-work lines of code using lambdas:
```
public string GetAggregateSortedListOfStatesUsingCsvRows()
{
     return GetUniqueSortedListOfStatesGivenCsvRows().Aggregate((a, b) => $"{a},{b}");
}
```

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,106 @@
...
+                throw new ArgumentNullException(nameof(filePath));
+            if (!File.Exists(filePath))
+                throw new FileNotFoundException();
 
-        // 3.
+            FilePath = filePath;
+        }
+
+        //--------------- 1 : CsvRows Property ----------------//
+        public IEnumerable<string> CsvRows
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Why not just return `File.ReadAllLines(FilePath)`, surely you don't need `yield return`?

File: Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,106 @@
...
+                    .OrderBy(p => p.Address.State)
+                    .ThenBy(p => p.Address.City)
+                    .ThenBy(p => p.Address.Zip);
+
+                return people;
+            }
+        }
+
+        //--------------- 5 : Filter By Email Address ----------------//
+        public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(Predicate<string> filter) =>
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Why the switch to standard query operators here?

File: Assignment/SampleData.cs [hunk 6]
```diff
@@ -1,30 +1,106 @@
...
+                .Aggregate((abbreviation1, abbreviation2) => $"{abbreviation1},{abbreviation2}");
+
+        //Static Parsing functions
+        public static Person ParsePerson(string row)
+        {
+            if (row is null)
+                throw new ArgumentNullException(nameof(row));
+
+            string[] values = row.Split(',');
+            string[] address = values.Skip(4).Take(4).ToArray();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Using seemingly arbitrary integers here is hard to read.  Consider naming the integers with enums or constants.

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,106 @@
...
+                .OrderBy(s => s)
+                .Distinct();
+
+        //--------------- 3 : Get Aggregated Sorted List Of States Using CSV Rows ----------------//
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            IEnumerable<string> uniqueSortedListOfStates = GetUniqueSortedListOfStatesGivenCsvRows();
+
+            string[] states = uniqueSortedListOfStates.ToArray();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Why use `.ToArray()` when 'string.Join()' takes an `IEnumerable<string>`?

**Comment by MarkMichaelis** [comment]
##  (mmwoodfo)

## Assignment8

| Build/Compile (Build Succeeded: 0 Error(s), 20 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5**


| Assignment  | **15** |
| --------------------- | ------ |
| 1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ❌ <br/>- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔ <br/> - Using LINQ, skip the first row in the `People.csv`. ✔ <br/> - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔  | ✔ |
| 2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔ <br/> Use `ISampleData.CsvRows` for your data source. ✔ <br/>- Don't forget the list should be unique. ✔ <br/>- Sort the list alphabetically ✔ <br/>- Include a test that leverages a hard coded list of Spokane based addresses. ✔ <br/>- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hard coded list). ✔ | ✔ |
| 3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` for your data source. ✔ <br/>- Given the array, consider using `string.Join` to combine the list into a single string. | ✔ |
| 4.Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects. ✔ <br/>- Use `ISampleData.CsvRows` as the source of the data. ✔ <br/>- Sort the list by State, City, Zip. (Sort the addresses first then select). ✔ <br/>- Be sure that `Person.Address` is also populated.  ✔ <br/>- Adding null validation to all the `Person` and `Address` properties is **optional**. <br/>- Consider using `ISampleData.CsvRows` in your test to verify your results. | ✔ |
| 5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔ <br/>- Use `ISampleData.People` for your data source. ✔ | ✔ |
| 6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use the `people` parameter for your data source. ✔ <br/>- At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔ <br/>- Don't forget the list should be unique. ✔ <br/>- It is recommended that at a minimum you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Appropriately handle all warnings | **❌** |
| Add the code analysis packages and appropriately handle all warnings* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment | Enable | Assignment8.sln |  |  | 4 |
| AssignmentTests | Enable | Assignment8.sln |  |  | 4 |


| Testing | **20** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=25, Passed=25, Failed=0, Skipped=0 | **95%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write awesome, well tested code that follows all the guidelines outlined in the book. | ✔ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance. | ✔ |
| Implement all homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `ISampleData`). Refactor both your SampleData and SampleDataAsync classes so there is minimal duplication.  This includes refactoring your tests so a significant amount of test code can be re-used.| ✔ |

**Pull request closed without merging**